### PR TITLE
lines can come with metadata

### LINKED ISSUE #2: Pass some extra data associated with each line through the matcher interface
For example we may be searching through lines in buffers and windows and want to remember the line number, bufnr, etc

I think a backwards compatible interface is to let the second argument (`lines`) be a list of tables, if it is then we do the scoring/matching on `line[0]` and then, if matched, return the whole table as is. I think we can also return the index even if the element is a string since that is probably the most commonly needed data

If that sounds okay I can submit a PR

### PR DESCRIPTION
**Pull request by IndianBoy42**
Fixes #2

### CODE REVIEW

File: lua/fuzzy_nvim/fzf_matcher.lua [hunk 1]
```diff
@@ -8,12 +8,11 @@ M.filter = function(_, pattern, lines, case_mode)
...
 
-	for _, line in ipairs(lines) do
+	for index, line_tbl in ipairs(lines) do
+		local line = line_tbl[1] or line_tbl
 		local score = fzf.get_score(line, pattern_obj, slab)
 		local positions = fzf.get_pos(line, pattern_obj, slab)
-		if (positions and #positions > 0) then
-			table.insert(ans, {line, positions, score})
-		end
+		if positions and #positions > 0 then table.insert(ans, { line_tbl, positions, score, index }) end
```
Comment (tzachar) [inline lua/fuzzy_nvim/fzf_matcher.lua]: Why change the formatting here?

Comment (IndianBoy42) [reply lua/fuzzy_nvim/fzf_matcher.lua]: sorry! stylua just autoformatted it

**Comment by IndianBoy42** [comment]
Is this okay?

**Comment by tzachar** [comment]
Can you also add the appropriate docs?

File: README.md [hunk 2]
```diff
@@ -23,25 +23,41 @@ use {'tzachar/fuzzy.nvim', requires = {'romgrk/fzy-lua-native'}}
...
 # Api
 
-This plugin supports a module with a single method: `filter`.
-The `filter` method receives three parameters: a pattern to match, a list of
-strings and an extra parameter passed to the fuzzy library: for fzy, its a
-boolean indicating `is_case_sensitive`, for fzf its the `case_mode` (see
-[this](https://github.com/nvim-telescope/telescope-fzf-native.nvim#lua-interface)). 
+This plugin supports a module with a single method: `filter`. The `filter`
+method receives three parameters: a pattern to match, a list of strings or
+tables where the first element is a string and an extra parameier passed to the
```
Comment (tzachar) [inline README.md]: parameier -> parameter

File: README.md [hunk 3]
```diff
@@ -23,25 +23,41 @@ use {'tzachar/fuzzy.nvim', requires = {'romgrk/fzy-lua-native'}}
...
+fuzzy library: for fzy, its a boolean indicating `is_case_sensitive`, for fzf
+its the `case_mode` (see
+[this](https://github.com/nvim-telescope/telescope-fzf-native.nvim#lua-interface)).
 
-The method returns a list of `{string, match_positions, score}`. For the
-meanings of match_positions and score see individual documentations of `fzf` and
-`fzy`.
+The method returns a list of `{string|table, match_positions, score, index}`.
+For the meanings of match_positions and score see individual documentations of
+`fzf` and `fzy`. `index` is the index of the matched string in the input list.
```
Comment (tzachar) [inline README.md]: not clear what 'input list' means here. Refer to the arguments of the function to make it clear.

File: README.md [hunk 4]
```diff
@@ -23,25 +23,41 @@ use {'tzachar/fuzzy.nvim', requires = {'romgrk/fzy-lua-native'}}
...
 local matcher = require('fuzzy_nvim')
 matcher:filter('abc', {'aabbcc', '123', 'a1b2c'})
 ```
 
-result with `fzf` is then: 
-`
-{ { "a1b2c", { 5, 3, 1 }, 58 }, { "aabbcc", { 5, 3, 1 }, 49 } }
-`
+result with `fzf` is then:
+`{ { "a1b2c", { 5, 3, 1 }, 58, 3 }, { "aabbcc", { 5, 3, 1 }, 49, 1 } }`
```
Comment (tzachar) [inline README.md:45]: should this be `{ "a1b2c", { 5, 3, 1 }, 58, 1 }`, or am I missing something?

**Pull request merged**